File: Array/ThreeNumberSum.py


# O(n^2) time | O(n) space
def ThreeNumberSum(arr,target):
	arr.sort()
	# print(arr,target)
	res=[]
	for i in range(len(arr)-2):
		k = arr[i]
		l = i+1
		r = len(arr)-1
		while(l<r):
			if(k+arr[l]+arr[r] == target):
				res.append([k,arr[l],arr[r]])
				l+=1
				r-=1
			elif (target < k+arr[l]+arr[r]):
				r=r-1
			elif(target > k+arr[l]+arr[r]):
				l = l+1
	return res


if __name__ == "__main__":
	print(ThreeNumberSum([-1,2,4,5,7,8,8],18))

File: Array/ThreeNumberSum.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;;

// o(n^2) | s(1)
vector<ll> ThreeSum(vector<ll> v,int target){

	sort(v.begin(), v.end());
	
	for(int i = 0;i < v.size()-2;i++){
		int j = i+1;
		int k = v.size()-1;
		while(j < v.size()){
			if((v[i]+v[j]+v[k]) == target){
				return  { v[i],v[j],v[k] };
			}else if((v[i]+v[j]+v[k]) > target){
				k--;
			}else{
				j++;
			}
		}
	}
	return {-100,-100,-100};
}


int main(){


	vector<ll> v = {12,3,1,2,-6,5,-8,6};
	ll t = 0;

	vector<ll> res = ThreeSum(v,t);
	cout<<res[0]<<" "<<res[1]<<" "<<res[2];
	return 0;

}



